### PR TITLE
update docs with coding standards

### PR DESCRIPTION
**Pull request by jroebu14**
Resolves NEWSWORLDSERVICE-1494

**Overall change:**
Adds agreed set of component standards to the docs.

Replaces old coding standards doc.

---

- [ ] I have assigned myself to this PR and the corresponding issues
- [ ] I have added the `cross-team` label to this PR if it requires visibility across World Service teams
- [ ] I have assigned this PR to the Simorgh project
- [ ] (BBC contributors only) This PR follows the [repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

**Testing:**

- [ ] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false yarn test:e2e:interactive`)
- [ ] This PR requires manual testing

### CODE REVIEW

File: docs/Coding-Standards/Clean-Code.md [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+
+  // Length of string
+  const length = data.length;
+
+  // Loop through every character in data
+  for (let i = 0; i < length; i++) {
+    // Get character code.
+    const char = data.charCodeAt(i);
+    // Make the hash
+    hash = (hash << 5) - hash + char;
```
Comment (andrewscfc) [inline docs/Coding-Standards/Clean-Code.md]: Not sure this is the best example as I had to look up this operator to understand what was going on (https://stackoverflow.com/questions/10422744/what-do-the-operators-double-arrow-and-single-pipe-mean-in-javascri) Imo this code mostly does need comments as most engineers won't encounter bit shifting operators in day-to-day coding 

Maybe change the 'good' example to explain the bit shifting op but leave out comments around getting the character code and looping through the data? That shows the obvious and less obvious stuff being commented appropriately.

File: docs/Coding-Standards/React-Components.md [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+### How?
+
+Put styled components in a file with the same filename as the component but with `.styled` before the `.jsx` file extension. For example, `index.styled.jsx`. Then reference the styled components in the component file.
+
+❌
+
+```jsx
+const StyledWrapper = styled.div`
+  display: grid;
```
Comment (andrewscfc) [inline docs/Coding-Standards/React-Components.md:148]: might want to change this example from a grid layout as it contradicts our other standard around the use of the grid layout for anything other than page layout

File: docs/Coding-Standards/React-Components.md [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+- Reusing the functionality common to all usages of a particular user experience from the Compound Component
+- Applying customisations for the specific usage at the page level; we are free to make changes here without worrying about reuse
+
+In our example these are `RelatedContentPromo` and `TopStoriesPromo` respectively. Differences to the Promos on the respective page types are applied in these Components and they are co-located with the Page.
+
+### Bringing it Together
+
+Below is the full worked example. The example shows our Article and Story pages implementing their own Applied Components for Related Content and Top Stories. They also use a Complex Component to implement a Byline at the top of the article, it needs no customisation between the page types. The Promo Compound Component and Byline Complex Component use the Primitive Components to present a resuable component for use across page types.
+
+![](https://paper-attachments.dropbox.com/s_2A43B8E3F9C1F972929EDC9878EAFF3FE87CC9300DB1C147F888F337C763ADF5_1656492576866_file.png)
```
Comment (andrewscfc) [inline docs/Coding-Standards/React-Components.md]: can these images be added to source control as these attachement links may not work long term? Also I can provide the draw.io original to source control with this doc:
[compound-components.drawio.zip](https://github.com/bbc/simorgh/files/9243214/compound-components.drawio.zip)

File: docs/Coding-Standards/React-Components.md [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+In the past we built, highly configurable, reusable components that required large amounts of props to be ‘drilled’ into component to allow them to work on a particular page; a particular example of this was the [StoryPromo](https://github.com/bbc/simorgh/blob/d025086b41fb8753ec1a3992977bef5b013410b7/src/app/containers/StoryPromo/index.jsx#L126) component. With these highly configurable components the complexity grew and grew as they were reused in more places.
+
+In the new approach we make simple highly reusable primitive components with few customisation options, we create CompoundComponents to guide reuse through a set of child components and then we apply these for our specific page types, pushing specialisation to the top of the component hierarchy.
+
+## Component Locations
+
+### Legacy Components
+
+Before looking to apply our component standards we have a huge body of components that will not meet the existing standards to provide a ‘clean slate’ these can be moved to a legacy location in the following manner:
```
Comment (andrewscfc) [inline docs/Coding-Standards/React-Components.md]: Should this be updated to be in the present tense now but record for posterity where they all came from?

File: docs/Coding-Standards/React-Components.md [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+
+## Typescript for Frontend Components
+
+- Our visual components should be written in typescript clearly document each component’s API and get benefits of compile-time type checking between components.
+- Types should be co-located with the component that uses them; they should only be maintained in a shared location where they are used by multiple visual components
+  - Here is an example of shared types in our BFF application: https://github.com/bbc/fabl-modules/tree/master/modules/application/simorgh-bff/src/models
+  - We can locate any shared types in `src/app/components/models`
+- Types should be as strict as possible
+  - Avoid use of the `any` type and try and define a more specific type
+  - Where a prop has a fixed set of options use a `[union](https://camchenry.com/blog/typescript-union-type)` [type](https://camchenry.com/blog/typescript-union-type) [](https://camchenry.com/blog/typescript-union-type)to restrict the prop to those fixed options
```
Comment (andrewscfc) [inline docs/Coding-Standards/React-Components.md:266]: We should comment explicity on the inability to use aliases here I think

File: docs/Coding-Standards/React-Components.md [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+## Readme
+
+All components should have a readme that includes
+
+- A plain text description of what the component does
+- API docs describing the component’s props
+- A11y notes: explaination of implmentation details specific to a11y requirements
+- Usage examples including RTL
+- List any third party dependencies in use
+- An explaination of any unsual/uninuitive implementation decisions required to work around an external bug (e.g. browser bug) or to deliver quickly to a deadline
```
Comment (andrewscfc) [inline docs/Coding-Standards/React-Components.md]:
```suggestion
- An explaination of any unsual/unintuitive implementation decisions required to work around an external bug (e.g. browser bug) or to deliver quickly to a deadline
```

File: docs/Coding-Standards/Styles.md [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+
+We should avoid cases like this and generally try to reduce our use of props for styling a component if it will be invoked multiple times on a page with different props.
+
+[CSS Grid Layout](https://developer.mozilla.org/en-US/docs/Web/CSS/grid) is a popular layout mechanism suitable for many use cases, unfortuntately, one of our supported browsers Opera Mini does [not support the API](https://caniuse.com/css-grid). Based on this we have the following guidelines
+
+## CSS Grid can only be used for page layout
+
+Where CSS Grid is used Opera Mini just fall back to the default inline layout so on a desktop this layout in css grid:
+
+![](https://paper-attachments.dropbox.com/s_2A43B8E3F9C1F972929EDC9878EAFF3FE87CC9300DB1C147F888F337C763ADF5_1657696680147_Screen+Shot+2022-07-05+at+09.57.12.png)
```
Comment (andrewscfc) [inline docs/Coding-Standards/Styles.md]: more images to add to source control

File: docs/Coding-Standards/Naming-Conventions.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+✅
+
+```jsx
+const TopicPromo = () => <div />;
+const useToggle = () => {};
+const getPath = () => {};
+```
+
+#### The definition name and filename should be identical. This makes it easier to locate code.
```
Comment (RetiredBlob) [inline docs/Coding-Standards/Naming-Conventions.md]:
```suggestion
#### The definition name and directory name should be identical. This makes it easier to locate code.
```

File: docs/Coding-Standards/Naming-Conventions.md [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+// TopicPromo/index.jsx
+const TopicPromo = () => <div />;
+
+// RadioSchedule/index.jsx
+const RadioSchedule = () => <div />;
+```
+
+#### Component test filenames should be identical to the component filename with `.test` preceding the file extension. This makes it easier to locate tests for the component.
+
+[**❌**](https://emojipedia.org/cross-mark/)
```
Comment (RetiredBlob) [inline docs/Coding-Standards/Naming-Conventions.md]:
```suggestion
❌
```

File: docs/Coding-Standards/Naming-Conventions.md [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+```
+
+✅
+
+```jsx
+TopicPromo; // index.js; // the component filename
+TopicPromo; // index.test.js; // the component test filename
+```
+
+#### Component and component test filenames should use the `.jsx` file extension. This makes it easier to identify which files in the codebase are components.
```
Comment (RetiredBlob) [inline docs/Coding-Standards/Naming-Conventions.md]:
```suggestion
#### Component and component test filenames should use the `.tsx` file extension. This makes it easier to identify which files in the codebase are components.
```

File: docs/Coding-Standards/Naming-Conventions.md [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+```jsx
+TopicPromo; // index.js;
+TopicPromo; // index.test.js;
+```
+
+✅
+
+```jsx
+TopicPromo; // index.jsx;
+TopicPromo; // index.test.jsx;
```
Comment (RetiredBlob) [inline docs/Coding-Standards/Naming-Conventions.md]:
```suggestion
TopicPromo; // index.tsx;
TopicPromo; // index.test.tsx;
```

File: docs/Coding-Standards/React-Components.md [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+
+## Keep styled components in a separate file
+
+### Why?
+
+Component files are easier to reason about when you separate the presentation from the content and behaviour aspects of a component. With this separation of concerns, there is less cognitive load when it comes to updating the content or behaviour of a component because everything concerning the presentation is out of sight and mind in a separate file.
+
+### How?
+
+Put styled components in a file with the same filename as the component but with `.styled` before the `.jsx` file extension. For example, `index.styled.jsx`. Then reference the styled components in the component file.
```
Comment (RetiredBlob) [inline docs/Coding-Standards/React-Components.md]:
```suggestion
Put styled components in a file with the same filename as the component but with `.styled` before the `.tsx` file extension. For example, `index.styled.jsx`. Then reference the styled components in the component file.
```

Comment (RetiredBlob) [inline docs/Coding-Standards/React-Components.md]:
```suggestion
Put styled components in a file with the same filename as the component but with `.styled` before the `.tsx` file extension. For example, `index.styled.jsx`. Then reference the styled components in the component file.
```

File: docs/Coding-Standards/Clean-Code.md [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+
+#### Always try to explain yourself in code
+
+TODO
+
+❌
+
+```js
+// add bad example
+```
```
Comment (RetiredBlob) [inline docs/Coding-Standards/Clean-Code.md]:
```suggestion
```js
// analytics endpoint
const URL = 'https://analytics.bbc.com';
```
```

File: docs/Coding-Standards/Clean-Code.md [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+
+✅
+
+```js
+// add good example
+```
```
Comment (RetiredBlob) [inline docs/Coding-Standards/Clean-Code.md]:
```suggestion
```js
const ANALYTICS_ENDPOINT = 'https://analytics.bbc.com';
```
```

File: docs/Coding-Standards/Clean-Code.md [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+❌
+
+```js
+// add bad example
+```
+
+✅
+
+```js
+// add good example
```
Comment (RetiredBlob) [inline docs/Coding-Standards/Clean-Code.md]:
```suggestion
// Changes here will need replicated to the content security policy
const CDN_URL = 'https://cdn.bbc.com';
```

File: docs/Coding-Standards/Clean-Code.md [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+```
+
+#### Use as warning of consequences
+
+TODO
+
+❌
+
+```js
+// add bad example
```
Comment (ExcitedBlob) [inline docs/Coding-Standards/Clean-Code.md:148]: // Don't change this
const CDN_URL = 'https://cdn.bbc.com';

Comment (RetiredBlob) [inline docs/Coding-Standards/Clean-Code.md]:
```suggestion
// Changing this is risky
const CDN_URL = 'https://cdn.bbc.com';
```

File: docs/Coding-Standards/Clean-Code.md [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+```
+
+✅
+
+```js
+// add good example
```
Comment (RetiredBlob) [inline docs/Coding-Standards/Clean-Code.md]:
```suggestion
// Tigrinya is misspelt in the API response - see [API-1234]
const mapServiceToProducer = (service) => {
  if (service === 'tigrinia') return 'tigrinya';
  return producer;
};
```

File: docs/Coding-Standards/Clean-Code.md [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
```
Comment (RetiredBlob) [inline docs/Coding-Standards/Clean-Code.md]:
```suggestion
// This is the detection method recommended by opera
// See https://dev.opera.com/articles/opera-mini-and-javascript/
const isOperaMini = () =>
  Object.prototype.toString.call(window.operamini) === '[object OperaMini]';
```

**Comment by RetiredBlob** [comment]
It's surprisingly hard to come up with bad examples

Screen Shot 2022-08-03 at 17 18 57

File: docs/Coding-Standards/Clean-Code.md [hunk 19]
```diff
@@ -0,0 +1,322 @@
...
+#### Write pure functions
+
+A function is called a pure function if it always returns the same result when given the same argument values and it has no side effects like modifying an argument or a global variable.
+
+Pure functions are also extremely independent meaning they are easy to move around, refactor, and reorganise in your code, making your applications more flexible and adaptable to future changes. It also makes them immune to bugs that have to do with shared mutable state.
+
+❌
+
+```js
+// add bad example
```
Comment (ExcitedBlob) [inline docs/Coding-Standards/Clean-Code.md:269]: // Modifies its argument
const getLastElement = data => data.pop();

// Accesses external variables, so not guaranteed to return the same value for the same argument
const getLastElement = () => data[data.length-1];

Comment (RetiredBlob) [inline docs/Coding-Standards/Clean-Code.md]:
```suggestion
// Modifies argument
const getLastElement = data => data.pop();

// Accesses external data, so not guaranteed to return the same result for the same arguments
const externalVariable = [1, 2, 3];
const getLastElement = () => externalVariable[externalVariable.length - 1]
```

File: docs/Coding-Standards/Clean-Code.md [hunk 20]
```diff
@@ -0,0 +1,322 @@
...
+```
+
+✅
+
+```js
+// add good example
```
Comment (RetiredBlob) [inline docs/Coding-Standards/Clean-Code.md]:
```suggestion
const getLastElement = data => data[data.length-1];
```

File: docs/Coding-Standards/React-Components.md [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+
+### Compound Components
+
+These components apply the ‘[Compound Component Pattern](https://kentcdodds.com/blog/compound-components-with-react-hooks)’ to create a ‘Compound Component’ that can be used to implement a particular user experience across different page types.
+
+In our worked example we create a `Promo` that allows customised versions of `Title` , `Timestamp` and `Image` to be used in different ways depending on the use case.
+
+### Complex Components
+
+These components are components that use Primitive components to build a reusable component that isn’t suitable to be implemented as a Compound Component. It needs limited or no customisation for usages on different pages. A Complex Component could later be converted to Compound Component if it needed significant customisation that differed across Page Types.
```
Comment (rebeccamcginn) [inline docs/Coding-Standards/React-Components.md]:
```suggestion
These components use Primitive components to build a reusable component that isn’t suitable to be implemented as a Compound Component. It needs limited or no customisation for usages on different pages. A Complex Component could later be converted to Compound Component if it needed significant customisation that differed across Page Types.
```

File: docs/Coding-Standards/React-Components.md [hunk 22]
```diff
@@ -0,0 +1,292 @@
...
+
+In our example below we have a `Byline` component used on both Article and Story Pages with no customisation at all.
+
+### Applied Compound Components
+
+At this level we take a ‘Compound Component’ and build an ‘Applied Compound Component’ that is very specific to the page’s use case. At this level we can apply specific styling and add in additional functionality completly bespoke to that page’s requirements; we are:
```
Comment (rebeccamcginn) [inline docs/Coding-Standards/React-Components.md]:
```suggestion
At this level we take a ‘Compound Component’ and build an ‘Applied Compound Component’ that is very specific to the page’s use case. At this level we can apply specific styling and add in additional functionality completely bespoke to that page’s requirements; we are:
```

File: docs/Coding-Standards/React-Components.md [hunk 23]
```diff
@@ -0,0 +1,292 @@
...
+
+### Bringing it Together
+
+Below is the full worked example. The example shows our Article and Story pages implementing their own Applied Components for Related Content and Top Stories. They also use a Complex Component to implement a Byline at the top of the article, it needs no customisation between the page types. The Promo Compound Component and Byline Complex Component use the Primitive Components to present a resuable component for use across page types.
+
+![](https://paper-attachments.dropbox.com/s_2A43B8E3F9C1F972929EDC9878EAFF3FE87CC9300DB1C147F888F337C763ADF5_1656492576866_file.png)
+
+### Summary
+
+In the past we built, highly configurable, reusable components that required large amounts of props to be ‘drilled’ into component to allow them to work on a particular page; a particular example of this was the [StoryPromo](https://github.com/bbc/simorgh/blob/d025086b41fb8753ec1a3992977bef5b013410b7/src/app/containers/StoryPromo/index.jsx#L126) component. With these highly configurable components the complexity grew and grew as they were reused in more places.
```
Comment (rebeccamcginn) [inline docs/Coding-Standards/React-Components.md]:
```suggestion
In the past we built highly configurable, reusable components that required large amounts of props to be ‘drilled’ into components to allow them to work on a particular page. An example of this was the [StoryPromo](https://github.com/bbc/simorgh/blob/d025086b41fb8753ec1a3992977bef5b013410b7/src/app/containers/StoryPromo/index.jsx#L126) component. With these highly configurable components the complexity grew and grew as they were reused in more places.
```

File: docs/Coding-Standards/React-Components.md [hunk 24]
```diff
@@ -0,0 +1,292 @@
...
+And all components meeting our standards will from now on will be maintained in the following directories:
+
+| Component Types                 | Location                    |
+| ------------------------------- | --------------------------- |
+| Primitive, Compound, Complex    | `src/app/components`        |
+| Applied Compound, Page Specific | `/src/app/components/pages` |
+
+### Applying Standardisation
+
+Where possible we want apply standardisation as an ‘opportunitistic refactor’. As we look to use existing components to deliver a feature, we commit to rebuilding those components or porting them across applying new best practices. Through doing this we gradually improve the consistency and standards of our components rather then embarking on a unrealistic drive to rebuild all components to a new standard. The necessity of delivering a feature keeps scope under control while the standards maintain the quality of the code for that feature.
```
Comment (rebeccamcginn) [inline docs/Coding-Standards/React-Components.md]:
```suggestion
Where possible we want to apply standardisation as an ‘opportunistic refactor’. As we look to use existing components to deliver a feature, we commit to rebuilding those components or porting them across applying new best practices. Through doing this we gradually improve the consistency and standards of our components rather then embarking on a unrealistic drive to rebuild all components to a new standard. The necessity of delivering a feature keeps scope under control while the standards maintain the quality of the code for that feature.
```

File: docs/Coding-Standards/React-Components.md [hunk 25]
```diff
@@ -0,0 +1,292 @@
...
+
+## Typescript for Frontend Components
+
+- Our visual components should be written in typescript clearly document each component’s API and get benefits of compile-time type checking between components.
```
Comment (rebeccamcginn) [inline docs/Coding-Standards/React-Components.md]:
```suggestion
- Our visual components should be written in TypeScript to clearly document each component’s API and get the benefits of compile-time type checking between components.
```

File: docs/Coding-Standards/React-Components.md [hunk 26]
```diff
@@ -0,0 +1,292 @@
...
+![](https://paper-attachments.dropbox.com/s_0A228B6278B754FD729F6444013C8CCE0645C344FE8C5686DEE4A9889366E8C3_1623683012926_Screenshot+2021-06-14+at+16.03.01.png)
+
+Component Health should include:
+
+- Link to screen reader UX
+- Accessibility acceptance criteria
+- Accessibility swarm document
+- Whether the component has ‘right-to-left’ support
+
+All docs linked too must be secure and not publically editable
```
Comment (rebeccamcginn) [inline docs/Coding-Standards/React-Components.md]:
```suggestion
All linked docs must be secure and not publicly editable
```

File: docs/Coding-Standards/React-Components.md (hunk 6 above)
Comment (rebeccamcginn) [inline docs/Coding-Standards/React-Components.md]:
```suggestion
- An explanation of any unusual/uninuitive implementation decisions required to work around an external bug (e.g. browser bug) or to deliver quickly to a deadline
```

File: docs/Coding-Standards/React-Components.md (hunk 2 above)
Comment (jroebu14) [reply docs/Coding-Standards/React-Components.md:148]: This example is a page layout. I can still change if it if you like though.

**Review by RetiredBlob — APPROVED** [review]

**Review by amoore108 — APPROVED** [review]
Very nice!

**Pull request merged**